src/coreclr/vm/i386/jithelp.S

```diff
@@ -387,16 +388,11 @@ C_FUNC(JIT_WriteBarrierEAX_Loc):
 LEAF_ENTRY JIT_WriteBarrier_Callable, _TEXT
     mov eax, edx
     mov edx, ecx
-    push    eax
-    call    1f
-1:
-    pop     eax
-2:
-.att_syntax
-    addl     $_GLOBAL_OFFSET_TABLE_+(2b-1b), %eax
-.intel_syntax noprefix
-    mov     eax, dword ptr [eax + C_FUNC(JIT_WriteBarrierEAX_Loc)@GOT]
+// Place JIT_WriteBarrierEAX address into stack to call it via ret instuction.
+    push eax
+    mov eax, dword ptr [C_FUNC(JIT_WriteBarrierEAX_Loc)]
     xchg    eax, dword ptr [esp]
+// Real return address is still on the stack at esp + 4.
     ret
```

Comment:
Are we sure that the above removed instructions are not trying to keep the hardware's call-ret stack balanced?

Reply:
If no IP-relative computation is needed I think the entire function can just be
```asm
mov eax, edx
mov edx, ecx
jmp dword ptr [C_FUNC(JIT_WriteBarrierEAX_Loc)]
```

Reply:
> Are we sure that the above removed instructions are not trying to keep the hardware's call-ret stack balanced?

Maybe I do not clearly understand the question. I am sure that the above removed instructions keep the hardware's call-ret info on stack balanced. Removed instructions just calculate wrong address.




> If no IP-relative computation is needed I think the entire function can just be
> 
> ```assembly
> mov eax, edx
> mov edx, ecx
> jmp dword ptr [C_FUNC(JIT_WriteBarrierEAX_Loc)]
> ```

Unfortunately this code jumps to `JIT_WriteBarrierEAX_Loc`, not to the address stored in `JIT_WriteBarrierEAX_Loc`:
```gdb
=> 0xf7606c71 <JIT_WriteBarrier_Callable>:      mov    eax,edx
   0xf7606c73 <JIT_WriteBarrier_Callable+2>:    mov    edx,ecx
   0xf7606c75 <JIT_WriteBarrier_Callable+4>:    jmp    0xf7ad49b4
gdb$ disassemble 0xf7ad49b4
Dump of assembler code for function JIT_WriteBarrierEAX_Loc:
   0xf7ad49b4:  enter  0x6075,0xf7
End of assembler dump.
gdb$ disassemble 0xf76075c8
Dump of assembler code for function JIT_WriteBarrierEAX:
   0xf76075c8 <+0>:     mov    DWORD PTR [edx],eax
   0xf76075ca <+2>:     cmp    eax,0xf1a0000c
   0xf76075d0 <+8>:     jb     0xf76075df <JIT_WriteBarrierEAX+23>
   0xf76075d2 <+10>:    shr    edx,0xa
   0xf76075d5 <+13>:    nop
   0xf76075d6 <+14>:    cmp    BYTE PTR [edx-0xa4007dc],0xff
   0xf76075dd <+21>:    jne    0xf76075e2 <JIT_WriteBarrierEAX+26>
   0xf76075df <+23>:    ret    
   0xf76075e0 <+24>:    nop
   0xf76075e1 <+25>:    nop
   0xf76075e2 <+26>:    mov    BYTE PTR [edx-0xa4007dc],0xff
   0xf76075e9 <+33>:    ret 
```

Reply:
> Maybe I do not clearly understand the question. I am sure that the above removed instructions keep the hardware's call-ret info on stack balanced. Removed instructions just calculate wrong address.

The hardware tracks `call`/`ret` pairs to be able to branch predict `ret` instructions better. It means that after this change the hardware might mispredict a lot of the upcoming `ret` instructions.

> Unfortunately this code jumps to JIT_WriteBarrierEAX_Loc, not to the address stored in JIT_WriteBarrierEAX_Loc:

That's odd. Maybe try AT&T syntax? It should assemble to `ff 25 imm32`.

Reply:
I can reproduce it with `clang-9 foo.S` locally. Using AT&T syntax instead works and assembles correctly:
```asm
.att_syntax
    jmpl *C_FUNC(JIT_WriteBarrierEAX_loc)
```

Reply:
https://bugs.llvm.org/show_bug.cgi?id=20511

Reply:
Thanks, AT&T syntax works.

Reply:
IP relative addressing is a requirement on Linux. While it works without it, it requires relocations when loading the code and LLVM linker even refuses to link such code with our current options. I had to fix it recently for Linux ARM.

Reply:
Seems `eax` dereferencing was missed in original IP-relative addressing. Updated a PR, @janvorli please take a look.

Reply:
Hmm, with GOT we do not need `JIT_WriteBarrierEAX_Loc` at all. The code below works properly:
```asm
    mov eax, edx
    mov edx, ecx
    push    eax
    call    1f
1:
    pop     eax
2:
.att_syntax
    addl     $_GLOBAL_OFFSET_TABLE_+(2b-1b), %eax
.intel_syntax noprefix
    mov     eax, dword ptr [eax + C_FUNC(JIT_WriteBarrierEAX)@GOT]
    xchg    eax, dword ptr [esp]
    ret
```

Reply:
I don't see how this would work with the W^X enabled. We don't want to call the JIT_WriteBarrierEAX in the executable, but rather its copy that is placed at a random memory location (in a dynamically allocated memory page).

Reply:
> I don't see how this would work with the W^X enabled. We don't want to call the JIT_WriteBarrierEAX in the executable, but rather its copy that is placed at a random memory location (in a dynamically allocated memory page).

Would current PR variant with JIT_WriteBarrierEAX_Loc dereferencing work with W^X enabled?

Reply:
Yes, if that works with W^X disabled, it would work with it enabled too. The JIT_WriteBarrierEAX_Loc points to the JIT_WriteBarrierEAX with W^X disabled and to its copy with W^X enabled.